src/rdkafka_sasl_oauthbearer_oidc.c

```diff
@@ -0,0 +1,529 @@
+        rd_free(new);
+        return buf;
+}
+
+
+/**
+ * @brief Implementation of Oauth/OIDC token refresh callback function,
+ *        will receive the JSON response after HTTP call to token provider,
+ *        then extract the jwt from the JSON response, and forward it to
+ *        the broker.
+ */
+void rd_kafka_oidc_token_refresh_cb(rd_kafka_t *rk,
+                                    const char *oauthbearer_config,
+                                    void *opaque) {
+        const int timeout_s = 20;
+        const int retry     = 4;
+        const int retry_ms  = 5 * 1000;
+
+        double exp;
+
+        cJSON *json = NULL;
+        cJSON *parsed_token, *payloads, *jwt_exp, *jwt_sub;
+
+        rd_http_error_t *herr;
+
+        char *jwt_token;
+        char *post_fields;
+        char *decoded_payloads = NULL;
+
+        struct curl_slist *headers = NULL;
+
+        const char *token_url;
+        const char *extension_str;
+        const char *sub;
+
+        size_t post_fields_size;
+        size_t extension_cnt;
+        size_t extension_key_value_cnt;
+
+        char set_token_errstr[512];
+
+        char **extensions;
+        char **extension_key_value;
+
+        if (rd_kafka_terminating(rk))
+                return;
+
+        rd_kafka_oidc_build_headers(rk->rk_conf.sasl.oauthbearer.client_id,
+                                    rk->rk_conf.sasl.oauthbearer.client_secret,
+                                    &headers);
+
+        /* Build post fields */
+        post_fields_size = strlen("grant_type=client_credentials&scope=") +
+                           strlen(rk->rk_conf.sasl.oauthbearer.scope) + 1;
+        post_fields = rd_malloc(post_fields_size);
+        rd_snprintf(post_fields, post_fields_size,
+                    "grant_type=client_credentials&scope=%s",
+                    rk->rk_conf.sasl.oauthbearer.scope);
+
+        token_url = rk->rk_conf.sasl.oauthbearer.token_endpoint_url;
+
+        herr = rd_http_post_expect_json(rk, token_url, headers, post_fields,
+                                        post_fields_size, timeout_s, retry,
+                                        retry_ms, &json);
+
+        if (unlikely(herr != NULL)) {
+                rd_kafka_log(rk, LOG_ERR, "OIDC",
+                             "Failed to retrieve OIDC "
+                             "token from \"%s\": %s (%d)",
+                             token_url, herr->errstr, herr->code);
+                rd_kafka_oauthbearer_set_token_failure(rk, herr->errstr);
+                rd_http_error_destroy(herr);
+                goto done;
+        }
+
+        if (!cJSON_HasObjectItem(json, "access_token")) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk,
+                    "Expected JSON response with "
+                    "\"access_token\" field");
+                goto done;
+        }
+
+        parsed_token = cJSON_GetObjectItem(json, "access_token");
```

Comment:
Can you call GetObjectItem() directly and check it for NULL? It voids the need for the separate HasObjectItem call above.